src/audio/drc/drc.c

```diff
@@ -0,0 +1,849 @@
+		}
+
+		comp_info(dev, "drc_cmd_set_data(), chunk size: %u msg_index %u",
+			  cdata->num_elems, cdata->msg_index);
+		dst = (unsigned char *)cd->config_new;
+		src = (unsigned char *)cdata->data->data;
+
+		/* Just copy the configuration. The EQ will be initialized in
+		 * prepare().
+		 */
+		ret = memcpy_s(dst + offset, size - offset, src,
+			       cdata->num_elems);
+		assert(!ret);
+
+		/* we can check data when elems_remaining == 0 */
+		if (cdata->elems_remaining == 0) {
+			/* The new configuration is OK to be applied */
+			cd->config_ready = true;
+
+			/* If component state is READY we can omit old
+			 * configuration immediately. When in playback/capture
+			 * the new configuration presence is checked in copy().
+			 */
+			if (dev->state == COMP_STATE_READY)
+				drc_free_parameters(&cd->config);
+
+			/* If there is no existing configuration the received
+			 * can be set to current immediately. It will be
+			 * applied in prepare() when streaming starts.
+			 */
+			if (!cd->config) {
+				cd->config = cd->config_new;
+				cd->config_new = NULL;
+			}
+		}
+		break;
+	default:
+		comp_err(dev, "drc_cmd_set_data(), invalid command");
+		ret = -EINVAL;
+		break;
+	}
+
+	return ret;
+}
+
+static int drc_cmd(struct comp_dev *dev, int cmd, void *data,
+		   int max_data_size)
+{
+	struct sof_ipc_ctrl_data *cdata = data;
+	int ret = 0;
+
+	comp_info(dev, "drc_cmd()");
+
+	switch (cmd) {
+	case COMP_CMD_SET_DATA:
+		ret = drc_cmd_set_data(dev, cdata);
+		break;
+	case COMP_CMD_GET_DATA:
+		ret = drc_cmd_get_data(dev, cdata, max_data_size);
+		break;
+	default:
+		comp_err(dev, "drc_cmd(), invalid command");
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+static int drc_trigger(struct comp_dev *dev, int cmd)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	comp_info(dev, "drc_trigger()");
+
+	return comp_set_state(dev, cmd);
+}
+
+static void drc_process(struct comp_dev *dev, struct comp_buffer *source,
+			struct comp_buffer *sink, int frames,
+			uint32_t source_bytes, uint32_t sink_bytes)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	buffer_invalidate(source, source_bytes);
+
+	cd->drc_func(dev, &source->stream, &sink->stream, frames);
+
+	buffer_writeback(sink, sink_bytes);
+
+	/* calc new free and available */
+	comp_update_buffer_consume(source, source_bytes);
+	comp_update_buffer_produce(sink, sink_bytes);
+}
+
+/* copy and process stream data from source to sink buffers */
+static int drc_copy(struct comp_dev *dev)
+{
+	struct comp_copy_limits cl;
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *sourceb;
+	struct comp_buffer *sinkb;
+	int ret;
+
+	comp_dbg(dev, "drc_copy()");
+
+	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer,
+				  sink_list);
+
+	/* Check for changed configuration */
+	if (cd->config_new) {
+		drc_free_parameters(&cd->config);
+		cd->config = cd->config_new;
+		cd->config_new = NULL;
+		ret = drc_setup(cd, sourceb->stream.channels, sourceb->stream.rate);
```

Comment:
i think we may have a memory leak here if we init the pre delay buffers mid stream, and im not sure if we want to reinit them. @singalsu thoughts?

Reply:
It should not have memory leak here. In drc_setup(), drc_reset_state() is called first which freed all pre-delay buffers.

Reply:
setup